_posts/2019-06-25-dynamic_linear_models_tfprobability/dynamic_linear_models_tfprobability.Rmd

```diff
@@ -0,0 +1,805 @@
+
+```{r}
+n_iterations <- 100
+n_param_samples <- 50
+n_forecast_samples <- 50
+
+optimizer <- tf$compat$v1$train$AdamOptimizer(0.1)
+
+fit_vi <-
+  function(ts,
+           ts_train,
+           model,
+           n_iterations,
+           n_param_samples,
+           n_forecast_steps,
+           n_forecast_samples) {
+    loss_and_dists <-
+      ts_train %>% sts_build_factored_variational_loss(model = model)
+    variational_loss <- loss_and_dists[[1]]
+    train_op <- optimizer$minimize(variational_loss)
+    
+    with (tf$Session() %as% sess,  {
+      sess$run(tf$compat$v1$global_variables_initializer())
+      for (step in 1:n_iterations) {
+        res <- sess$run(train_op)
+        loss <- sess$run(variational_loss)
+        if (step %% 1 == 0)
+          cat("Loss: ", as.numeric(loss), "\n")
+      }
+      variational_distributions <- loss_and_dists[[2]]
+      posterior_samples <-
+        Map(
+          function(d)
+            d %>% tfd_sample(n_param_samples),
+          variational_distributions %>% reticulate::py_to_r() %>% unname()
+        )
+      forecast_dists <-
+        ts_train %>% sts_forecast(model, posterior_samples, n_forecast_steps)
+      fc_means <- forecast_dists %>% tfd_mean()
+      fc_sds <- forecast_dists %>% tfd_stddev()
+      
+      c(posterior_samples,
+        fc_means,
+        fc_sds) %<-%
+        sess$run(list(posterior_samples,
+                      fc_means,
+                      fc_sds))
+    })
+    
+    list(variational_distributions,
+         posterior_samples,
+         fc_means[, 1],
+         fc_sds[, 1])
+  }
+
+c(param_distributions,
+  param_samples,
+  fc_means,
+  fc_sds) %<-% fit_vi(
+    ts,
+    ts_train,
+    model,
+    n_iterations,
+    n_param_samples,
+    n_forecast_steps,
+    n_forecast_samples
+  )
+
+```
+
+For brevity, we haven't computed smoothed and/or filtered estimates for the overall model. In this example, this being a _sum_ model, we want to show something else instead: the way it decomposes into components.
+But first, the forecasts:
+
+```{r}
+forecast_df <- df %>%
+  add_column(pred_mean = c(rep(NA, length(ts_train)), fc_means)) %>%
+  add_column(pred_sd = c(rep(NA, length(ts_train)), fc_sds))
+
+
+ggplot(forecast_df, aes(x = month, y = n)) +
+  geom_line(color = "grey") +
+  geom_line(aes(y = pred_mean), color = "cyan") +
+  geom_ribbon(
+    aes(ymin = pred_mean - 2 * pred_sd, ymax = pred_mean + 2 * pred_sd),
+    alpha = 0.2,
+    fill = "cyan"
+  ) +
+  theme(axis.title = element_blank())
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "AirPassengers, 12-months-ahead forecast."}
+knitr::include_graphics("images/airp_forecast.png")
+```
+
+A call to [sts_decompose_by_component](https://rstudio.github.io/tfprobability/reference/sts_decompose_by_component.html) yields the -- centered -- components, a linear trend and a seasonal factor:
```

Comment:
Are the em-dashes here to emphasize centered? If so, you might italicize or embolden.

Reply:
ehm, not really... more to tell people (yeah, that's on purpose / not an error ;-))
I guess then parentheses should be more appropriate